Make statx available for all Linux targets

alyssais opened the pull request:
I _think_ all this should be available on all Linux targets now, but I'm not very familiar with uClibc, so I'm not sure about it.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @Amanieu (or someone else) soon.

Please see [the contribution instructions](https://github.com/rust-lang/libc/blob/master/CONTRIBUTING.md) for more information.

alyssais commented:
Oh, is there no musl/uClibc CI?

Amanieu commented:
We do have it but it is only run by bors, not the normal CI for PRs.

@bors r+

bors commented:
:pushpin: Commit d37715dd8a84fb203f7299cdc0b1f14d76997274 has been approved by `Amanieu`

<!-- @bors r=Amanieu d37715dd8a84fb203f7299cdc0b1f14d76997274 -->
<!-- homu: {"type":"Approved","sha":"d37715dd8a84fb203f7299cdc0b1f14d76997274","approver":"Amanieu"} -->

bors commented:
:hourglass: Testing commit d37715dd8a84fb203f7299cdc0b1f14d76997274 with merge 32413d6f3c17e11e0226c6fae3c50e73b19c32ff...
<!-- homu: {"type":"BuildStarted","head_sha":"d37715dd8a84fb203f7299cdc0b1f14d76997274","merge_sha":"32413d6f3c17e11e0226c6fae3c50e73b19c32ff"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/libc/runs/5457895302?check_suite_focus=true)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/libc/runs/5457895302?check_suite_focus=true","builder_name":"checks-actions"} -->

Amanieu commented:
You need to change libc-test/build.rs to include the statx header file in the C tests.

JohnTitor commented:
Ping @alyssais, could you address the above comment?

bors commented:
:umbrella: The latest upstream changes (presumably #2808) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

germag commented on src/unix/linux_like/linux/mod.rs:
```diff
@@ -3431,6 +3490,14 @@ extern "C" {
 
     pub fn shm_open(name: *const c_char, oflag: ::c_int, mode: mode_t) -> ::c_int;
 
+    pub fn statx(
+        dirfd: ::c_int,
+        pathname: *const c_char,
+        flags: ::c_int,
+        mask: ::c_uint,
+        statxbuf: *mut statx,
+    ) -> ::c_int;
+
```
AFIAK the current version of musl (1.2.3) doesn't implement a wrapper for`SYS_statx`.
But, I think it's worth leaving the rest (struct statx, constant, etc) available for all linux target

alyssais replied:
My mistake, you're correct.

germag commented:
I'm quite interested in this PR

alyssais commented:
I've rebased the PR and updated it to drop the statx function, which as @germag pointed out, I was wrong about being present on Musl.  Hopefully this also avoids the need to add a header to build.rs, although I'm not sure.

alyssais commented:
Thinking about it some more, is it right to expose even the structs if Musl doesn't? I thought it did, but it turns out the `struct statx` in Musl is internal-only, so to use it in a C program one would have to import <linux/stat.h>.  What if Musl came up with its own statx API that used different structs?

germag commented:
sorry for stepping in, are there any additional problems with this PR

JohnTitor commented:
> You need to change libc-test/build.rs to include the statx header file in the C tests.

The diff doesn't include the change mentioned above, have you tested that libc-test passes?

bors commented:
:umbrella: The latest upstream changes (presumably #2941) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

alyssais commented:
> > Thinking about it some more, is it right to expose even the structs if Musl doesn't? I thought it did, but it turns out the `struct statx` in Musl is internal-only, so to use it in a C program one would have to import <linux/stat.h>. What if Musl came up with its own statx API that used different structs?
> 
> if that were to happen, which I don't think it would, they would be incompatible with both the glibc and linux, since the `statx` struct is defined in the kernel headers.

I followed up on this in the #musl IRC channel, and the consensus there was that there's no guarantee that kernel headers and libc headers will be compatible for the same functionality, and that the libc crate probably shouldn't expose interfaces that come straight from the kernel but are not exposed by libc.

If libc does this in other places, then maybe it's okay, but I'd really like to hear the thoughts of a libc crate maintainer on this specific issue before I go further with this, since I'm not confident it's the way to go.

workingjubilee commented:
Mmm. I don't think we should be exposing this for all Linux targets. Mister Musl is correct, here. For instance, there's `AT_{ElfAuxval}` constants that are exposed by musl and glibc on a target-neutral basis, but that are _not_ defined in a target-neutral way by the kernel. Those libcs have made their own commitments, effectively, on how to make that work going forward, in the (admittedly unlikely) event the kernel introduces a new definition which clashes. I believe it's okay for us to "round up" a bit for certain definitions, but the lack of support from a notable and widely-used libc makes me hesitate. Fixing a struct definition can be easier than fixing a function definition, but it is definitely still annoying to do and liable to be highly breaking.

JohnTitor commented:
Triage: The arguments above make some sense to me. Let's close this PR for now and continue the discussion if someone wants to support it strongly. Thanks @alyssais for the PR anyway!
